### PR TITLE
update work status condition when recreate failed

### LINKED ISSUE #4282: After a resource is deleted, the resource fails to be delivered again. The work corresponding to the resource is not in the synchronization failure state.
**What happened**:
I created a service on the federation and successfully delivered a cluster. I limit the number of services in the default  namespace to 0. I deleted the service from the member cluster. Failed to deliver the service.
image

**What you expected to happen**:
Controller synchronization service status
**How to reproduce it (as minimally and precisely as possible)**:
Create a service on the federation and successfully delivered a cluster.
Occupy the port configured for the service
Delete the service from the cluster.
Failed to deliver the federated data again.
**Anything else we need to know?**:

**Environment**:
- Karmada version: 1.8.0-preview1
- kubectl-karmada or karmadactl version (the result of `kubectl-karmada version` or `karmadactl version`):
- Others:

### PR DESCRIPTION
**Pull request by zhzhuang-zju**
**What type of PR is this?**
/kind feature

**What this PR does / why we need it**:
When a member cluster deletes a resource and the control plane fails to re-create it, the event needs to be recorded and the status of the work needs to be refreshed.
**Which issue(s) this PR fixes**:
Fixes #4282

**Special notes for your reviewer**:
NONE
**Does this PR introduce a user-facing change?**:

```release-note
`karmada-controller-manager`: Fixed a corner case that `applied` status on Work/ResourceBinding not updated in case of re-create failed.
```

### CODE REVIEW

**Comment by zhzhuang-zju** [comment]
Verification report:
1. follow https://github.com/karmada-io/karmada/issues/4282#issuecomment-1829263012 to reproduce 
2. check work status
```bash
➜  karmada git:(master) ✗ ka get work -A
NAMESPACE            NAME                              APPLIED   AGE
karmada-es-member1   my-service-7ddc6b975              False     51s
➜  karmada git:(master) ✗ ka get work my-service-7ddc6b975 -n karmada-es-member1 -oyaml
...
status:
  conditions:
  - lastTransitionTime: "2023-12-01T01:34:48Z"
    message: 'services "my-service" is forbidden: exceeded quota: compute-resources,
      requested: services=1, used: services=1, limited: services=0'
    reason: ReCreateFailed
    status: "False"
    type: Applied
  manifestStatuses:
  - health: Unhealthy
    identifier:
      kind: Service
      name: my-service
      namespace: default
      ordinal: 0
      resource: ""
      version: v1
    status:
      loadBalancer: {}
➜  karmada git:(master) ✗ ka describe Service my-service                               
...
Events:
  Type     Reason                  Age                    From                                Message
  ----     ------                  ----                   ----                                -------
  Warning  SyncFailed              2m40s (x3 over 2m40s)  execution-controller                Failed to create or update resource(default/my-service) in member cluster(member2): the informer of cluster(member2) has not been initialized
  Normal   SyncSucceed             2m40s                  execution-controller                Successfully applied resource(default/my-service) to cluster member2
  Normal   ScheduleBindingSucceed  82s (x16 over 22h)     default-scheduler                   Binding has been scheduled successfully.
  Normal   SyncWorkSucceed         82s (x3 over 2m40s)    binding-controller                  Sync work of resourceBinding(default/my-service-service) successful.
  Normal   ApplyPolicySucceed      82s                    resource-detector                   Apply policy(default/foo) succeed
  Normal   SyncSucceed             82s                    execution-controller                Successfully applied resource(default/my-service) to cluster member1
  Normal   AggregateStatusSucceed  38s (x6 over 2m40s)    resource-binding-status-controller  Update resourceBinding(default/my-service-service) with AggregatedStatus successfully.
  Warning  SyncFailed              18s (x13 over 38s)     work-status-controller              Failed to create or update resource(default/my-service) in member cluster(member1): services "my-service" is forbidden: exceeded quota: compute-resources, requested: services=1, used: services=1, limited: services=0
```
3. Increase the resourceQuota of cluster one
```bash
#resourceQuota10.yaml
apiVersion: v1
kind: ResourceQuota
metadata:
  name: compute-resources
spec:
  hard:
    services: "10"
```
4. After the control plane successfully recreates the resource, check the work status again
```bash
➜  karmada git:(master) ✗ ka get work -A
NAMESPACE            NAME                              APPLIED   AGE
karmada-es-member1   my-service-7ddc6b975              True      2m10s
➜  karmada git:(master) ✗ ka get work my-service-7ddc6b975 -n karmada-es-member1 -oyaml
...
status:
  conditions:
  - lastTransitionTime: "2023-12-01T01:36:10Z"
    message: Manifest has been successfully applied
    reason: ReCreateSuccessful
    status: "True"
    type: Applied
  manifestStatuses:
  - health: Unhealthy
    identifier:
      kind: Service
      name: my-service
      namespace: default
      ordinal: 0
      resource: ""
      version: v1
    status:
      loadBalancer: {}
➜  karmada git:(master) ✗ ka describe Service my-service                               
...
Events:
  Type     Reason                  Age                    From                                Message
  ----     ------                  ----                   ----                                -------
  Warning  SyncFailed              3m52s (x3 over 3m52s)  execution-controller                Failed to create or update resource(default/my-service) in member cluster(member2): the informer of cluster(member2) has not been initialized
  Normal   SyncSucceed             3m52s                  execution-controller                Successfully applied resource(default/my-service) to cluster member2
  Normal   ScheduleBindingSucceed  2m34s (x16 over 22h)   default-scheduler                   Binding has been scheduled successfully.
  Normal   SyncWorkSucceed         2m34s (x3 over 3m52s)  binding-controller                  Sync work of resourceBinding(default/my-service-service) successful.
  Normal   ApplyPolicySucceed      2m34s                  resource-detector                   Apply policy(default/foo) succeed
  Normal   SyncSucceed             2m34s                  execution-controller                Successfully applied resource(default/my-service) to cluster member1
  Warning  SyncFailed              69s (x14 over 110s)    work-status-controller              Failed to create or update resource(default/my-service) in member cluster(member1): services "my-service" is forbidden: exceeded quota: compute-resources, requested: services=1, used: services=1, limited: services=0
  Normal   AggregateStatusSucceed  28s (x7 over 3m52s)    resource-binding-status-controller  Update resourceBinding(default/my-service-service) with AggregatedStatus successfully.
  Normal   SyncSucceed             28s                    work-status-controller              Successfully applied resource(default/my-service) to cluster member1

```

**Comment by zhzhuang-zju** [comment]
cc @CuiDengdeng @RainbowMango

File: pkg/controllers/status/work_status_controller.go [hunk 1]
```diff
@@ -280,7 +280,13 @@ func (c *WorkStatusController) handleDeleteEvent(key keys.FederatedKey) error {
 		return nil
 	}
 
-	return c.recreateResourceIfNeeded(work, key)
+	reCreateErr := c.recreateResourceIfNeeded(work, key)
+	if reCreateErr != nil {
+		c.updateAppliedCondition(work, metav1.ConditionFalse, "ReCreateFailed", reCreateErr.Error())
```
Comment (chaosi-zju) [inline pkg/controllers/status/work_status_controller.go:285]: `updateAppliedCondition` has a `error` type return value

File: pkg/controllers/status/work_status_controller.go [hunk 2]
```diff
@@ -280,7 +280,13 @@ func (c *WorkStatusController) handleDeleteEvent(key keys.FederatedKey) error {
...
+		return reCreateErr
+	}
+	c.updateAppliedCondition(work, metav1.ConditionTrue, "ReCreateSuccessful", "Manifest has been successfully applied")
```
Comment (chaosi-zju) [inline pkg/controllers/status/work_status_controller.go:288]: ditto

File: pkg/controllers/status/work_status_controller.go [hunk 3]
```diff
@@ -184,7 +184,7 @@ func (c *WorkStatusController) syncWorkStatus(key util.QueueKey) error {
 	fedKey, ok := key.(keys.FederatedKey)
 	if !ok {
 		klog.Errorf("Failed to sync status as invalid key: %v", key)
-		return fmt.Errorf("invalid key")
+		return nil
```
[the remark this answers is not in the document]
Comment (zhzhuang-zju) [reply pkg/controllers/status/work_status_controller.go:187]: @chaosi-zju @XiShanYongYe-Chang Please take a look and see if the modifications here are reasonable

[the remark this answers is not in the document]
Comment (RainbowMango) [reply pkg/controllers/status/work_status_controller.go:187]: > If the input parameter key is not the type we expect, then return error and then re-enter the queue.

Can you remind me when this situation would happen?

[the remark this answers is not in the document]
Comment (zhzhuang-zju) [reply pkg/controllers/status/work_status_controller.go:187]: > Can you remind me when this situation( I mean invalid key) would happen?

I didn't consider a specific scenario, just thinking from a code perspective.

> Is this change required to fix this issue?

It actually doesn't, it's unrelated to this issue.

[the remark this answers is not in the document]
Comment (RainbowMango) [reply pkg/controllers/status/work_status_controller.go:187]: Please revert this change if it's unrelated. As this patch is trying to fix things, we should make minimal changes.

**Review by RainbowMango — APPROVED** [review]
/lgtm
/approve

I updated the release notes.
Given this is a corner case and not a serious issue, we can skip backport.

**Pull request merged**